JPA Step2: 연관 관계 매핑 리뷰 요청 드립니다.

programmer-sjk opened the pull request:
안녕하세요 원철님.

지난번 피드백 주신 두가지 사항에 대해 전달드립니다.

**피드백 1: 프로퍼티 옵션**
```
spring.jpa.show-sql=true
spring.jpa.properties.hibernate.format_sql=true
```
주신 링크를 통해 `show-sql` 옵션을 `logging.level.org.hibernate.SQL=DEBUG` 옵션으로 수정했습니다.
`spring.jpa.properties.hibernate.format_sql=true` 옵션은 쿼리를 보는 가독성이 너무 달라져서 그대로 유지했습니다.

**피드백 2: 테스트 코드 비교 방식**
  주신 의견은 `assertThat(expect == result).isTrue();` 방식 보다는  `assertThat(expect).isEqualTo(result)` 방식은 어떤지 피드백을 주셨는데요. 개인적으로 팀 컨벤이션이나 취향차이로 갈릴 것 이라 생각했습니다. 그러다가 `assertEquals` 메소드가 있는걸 알게 됐고 이게 더 심플하다는 생각이 들어  `assertEquals(expect, result)` 방식으로 변경했습니다.

programmer-sjk committed:
docs(guide): README.md 요구사항 작성

programmer-sjk committed:
feat(domain):연관관계 맵핑

programmer-sjk committed:
test(fixture): 테스트용 fixture 함수 추가

programmer-sjk committed:
test(repository): repository 테스트 수정

programmer-sjk committed:
feat(domain): 연관관계에 fetch, joinColumn 추가

programmer-sjk committed:
chore(properties): 피드백 반영하여 logging.level.org.hibernate.SQL=DEBUG 옵션 추가

programmer-sjk committed:
test: assertEquals 메소드 적용

programmer-sjk committed:
refactor(domain): ForeignKey name 추가

programmer-sjk committed:
test: 변수명 수정

heowc commented on README.md:
```diff
@@ -55,3 +55,75 @@ create table user
+
+## 2단계 - 연관 관계 매핑
+
+### 요구 사항
```
요구사항을 잘 정리해주셨네요 👍

heowc commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -8,11 +8,14 @@
     private Long id;
-    private Long writerId;
-    private Long questionId;
+    @ManyToOne(fetch = FetchType.LAZY)
```
`LAZY`로 해주셨네요 👍 
혹시 이유를 알 수 있을까요?

추가로 관련해서 LAZY로 동작하는지 테스트 케이스를 작성해주시면 더 좋을 것 같습니다 (굳이 작성 안하셔도 괜찮습니다 👍 )

heowc commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -76,8 +79,8 @@ public void setDeleted(boolean deleted) {
                 "id=" + id +
-                ", writerId=" + writerId +
-                ", questionId=" + questionId +
+                ", writer=" + writer +
```
`toString`에서 Entity와 연관관계를 맺고 있는 객체를 그대로 넣는건 지양하는 것이 좋다고 생각하는데요.
어떻게 생각하시나요? 🤔

heowc commented on src/test/java/qna/fixture/TestAnswerFactory.java:
```diff
@@ -0,0 +1,11 @@
+import qna.domain.Question;
+import qna.domain.User;
+
+public class TestAnswerFactory {
```
테스트 작성에 용이하도록 Factory 클래스를 만들어주셨네요 👍 
관련 클래스들이 생기면서 AnswerTest 등 기존 클래스들이 사용되지 않는 것 같은데 확인 부탁드릴게요 🙇

heowc commented on src/test/java/qna/fixture/TestDeleteHistoryFactory.java:
```diff
@@ -0,0 +1,13 @@
+
+public class TestDeleteHistoryFactory {
+    public static DeleteHistory create(User deleteBy) {
+        return new DeleteHistory(ContentType.QUESTION,1L , deleteBy, LocalDateTime.now());
```
```suggestion
        return new DeleteHistory(ContentType.QUESTION, 1L , deleteBy, LocalDateTime.now());
```

heowc reviewed (changes requested):
2단계 잘 작성해주셨습니다 👍 

피드백 관련해서 많이 고민해보시고 상세히 작성해주셔서 제가 다 뿌듯한 것 같습니다 💯 💯
2번은 말씀대로 컨벤션일 수 있기 때문에 저도 가볍게 제안만 드려봤어요 😄

--- 

해당 단계 관련해서 소소한 피드백을 드렸습니다 🙇 
궁금하신 점이나 의논할 것이 있다면 깃헙 코멘트나 슬랙 DM 으로 연락주세요 🙏

programmer-sjk committed:
refactor: 코드 리뷰 반영

전체 코드 리뷰 반영
- fixture로 사용되지 않는 변수 제거
- @DataJpaTest에 쿼리 보이지 않느 설정 추가 (중복 이슈)
- Lazy Loading 테스트 추가

programmer-sjk commented:
안녕하세요 원철님 리뷰 부탁드립니다!! 

**피드백 1**
```
LAZY로 해주셨네요 👍
혹시 이유를 알 수 있을까요?

추가로 관련해서 LAZY로 동작하는지 테스트 케이스를 작성해주시면 더 좋을 것 같습니다 (굳이 작성 안하셔도 괜찮습니다 👍 )
```
-> default가 `Eager` 이기 때문에 `N+1` 문제를 방지하고자 넣었습니다. 사실 이번 미션에서는 넣지 않아도 되는 설정이지만
미션임과 동시에 학습의 기회이기 때문에 추가해 보았습니다.
-> JPA를 처음 해봐서 그런지 테스트 코드로 확인하는데 많이 삽질을 했고 다르게 표현하면 배웠네요 ㅋㅋ 
`AnswerRepositoryTest`에 테스트 코드 추가 했는데 한 번 보시고 피드백 있으면 남겨주세요

**피드백 2**
```
toString에서 Entity와 연관관계를 맺고 있는 객체를 그대로 넣는건 지양하는 것이 좋다고 생각하는데요.
어떻게 생각하시나요?
```
-> 검색해보니 `순환 참조` 이슈가 있을 수 있네요. 생각치 못한 내용인데 원철님 덕분에 알게 되었습니다. 수정했습니다!

**피드백 3**
```
테스트 작성에 용이하도록 Factory 클래스를 만들어주셨네요 👍
관련 클래스들이 생기면서 AnswerTest 등 기존 클래스들이 사용되지 않는 것 같은데 확인 부탁드릴게요 🙇
```
-> 사용하지 않는 기존 테스트 변수들 삭제했습니다 😄

programmer-sjk committed:
test: 불필요한 코드 제거

heowc commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -57,21 +86,41 @@ void findDeletedById() {
         assertThat(result).hasSize(0);
     }
+
+    @DisplayName("답변 조회시 writer, question이 지연로딩 되는지 확인한다")
```
지연로딩 테스트 👍👍

heowc commented on src/test/java/qna/fixture/TestUserFactory.java:
```diff
@@ -0,0 +1,9 @@
+
+import qna.domain.User;
+
+public class TestUserFactory {
```
static만 있는 클래스라면 기본 생성자를 막는건 어떨까요? 🤔

heowc reviewed (approved):
피드백까지 빠르게 잘 반영해주셨습니다 👍

다음단계를 진행해주시면 될 것 같아요 🙇

heowc commented:
> 안녕하세요 원철님 리뷰 부탁드립니다!! 
> 
> 
> 
> **피드백 1**
> 
> ```
> 
> LAZY로 해주셨네요 👍
> 
> 혹시 이유를 알 수 있을까요?
> 
> 
> 
> 추가로 관련해서 LAZY로 동작하는지 테스트 케이스를 작성해주시면 더 좋을 것 같습니다 (굳이 작성 안하셔도 괜찮습니다 👍 )
> 
> ```
> 
> -> default가 `Eager` 이기 때문에 `N+1` 문제를 방지하고자 넣었습니다. 사실 이번 미션에서는 넣지 않아도 되는 설정이지만
> 

eager가 즉시로딩이고 lazy가 지연로딩이기 때문에 오히려 N+1 문제는 lazy 설정시 발생합니다 😅

여쭤본 이유는 각각의 장단점이 있기 때문에 알고 쓰셨는지 궁금했어요~ 혹시 잘 못 알고 계셨거나 부족한 부분이 있다면 추가로 학습해보시면 좋을 것 같습니다 👍

궁금하신게 있으시면 언제근 연락주세요 🙏
